### ISSUE TITLE
Tiles in large maps should excluded from rendering when they are drawn outside the camera's view

### ISSUE THREAD

**Issue by craftworkgames** [craftworkgames/MonoGame.Extended#76]
@emersont1 on [gitter](https://gitter.im/craftworkgames/MonoGame.Extended) asked 

> if and how culling is done with the TMX support?

I took a look at the code and noticed that it's not doing any optimization in this area. I suspect rendering performance would degrade significantly with large maps. It should be relatively simple to only render tiles that are inside the camera's bounding rectangle.

This looks the relevant code https://github.com/craftworkgames/MonoGame.Extended/blob/master/Source/MonoGame.Extended/Maps/Tiled/TiledTileLayer.cs#L142

I guess those for loops should take rectangle as a parameter and only render tiles that fall inside, or partially overlap the rectangle bounds.

**Comment by cupstudio** [comment]
I was just going to ask this question.  I wrote my own quick and dirty tile system, but I'd like to switch to monogame.extended and am planning to make some changes to my code so that it can incorporate Extended as well as the Draw/Update bounds.

In my initial version, each layer is just two 2D Arrays.  One for the Tile ID, the other for a Vector2 position of the tile (which is created once in construction with number of rows, columns and tilewidth for the layer - it is immutable).  Everything else can be inferred from those two pieces of data.    At that point you can just give the position of the camera (which may be better to do in the Update phase depending on what else you are doing in your draw - and you can cache the position so re-calculations aren't necessary if there is no movement).

Here was my initial proof of concept code.   This ran at a smooth 60 fps ( my FPS class was updating in the Draw() method to ensure to get an actual average of draw calls) with 160,000 tiles (400x400).  I don't think there would be much of an upper limit though - the math is the same regardless because there is no enumeration.  The same code can be used to only update entities within a similar (though probably larger) range.


        public MapCoordinates GetStartingLocation(Vector2 location)
        {
            int row;
            int col;


            if (location.X < MapStartX)
            {
                col = 0;
            }
            else
            {
                if (location.X + SvcProvider.ScreenWidth > MapWidth)
                {
                    // -1 to account for 0-based index
                    col = MapColumns - 1;
                }
                else
                {
                    col = (int)Math.Floor(Math.Abs(MapStartX - location.X) / TileWidth);
                    if (col >= MapColumns)
                    {
                        col = MapColumns - 1;
                    }
                }
            }

            if (location.Y < MapStartY)
            {
                row = 0;
            }
            else
            {
                if (location.Y + SvcProvider.ScreenHeight >= MapHeight)
                {
                    // -1 to account for 0-based index
                    row = MapRows - 1;
                }
                else
                {
                    row = (int)Math.Floor(Math.Abs(MapStartY - location.Y) / TileHeight);
                    if (row >= MapRows)
                    {
                        row = MapRows - 1;
                    }
                }
            }

            return new MapCoordinates(row, col);
        }



And the layer's Draw method:

        /// <summary>
        /// Draw the tiles in this layer, starting with row, col
        /// </summary>
        /// <param name="spriteBatch"></param>
        /// <param name="row">Row to start at</param>
        /// <param name="col">Column to start at</param>
        public void DrawTiles(SpriteBatch spriteBatch, int row, int col)
        {
            // calculate rows / cols per screen
            int colsPerScreen = SvcProvider.ScreenWidth / tileSet.TileWidth;
            int rowsPerScreen = SvcProvider.ScreenHeight / tileSet.TileHeight;

            // add some padding to the number of rows and columns
            colsPerScreen++;
            rowsPerScreen += 2;

            int totalRows = Tiles.GetLength(0);
            int totalCols = Tiles.GetLength(1);

            // If we are near the edge of the map
            if (totalRows - row < rowsPerScreen)
            {
                rowsPerScreen = Math.Abs(totalRows - row);
            }
            if (totalCols - col < colsPerScreen)
            {
                colsPerScreen = Math.Abs(totalCols - col);
            }

            // Draw Tiles
            int currentRow = row;
            int currentCol = col;
            for (int x = 0; x < rowsPerScreen; x++)
            {
                if (x > 0)
                {
                    currentRow++;
                    currentCol = col;
                }
                for (int y = 0; y < colsPerScreen; y++)
                {
                    if (y > 0)
                    {
                        currentCol++;
                    }
                    DrawTile(spriteBatch, currentRow, currentCol);
                }
            }
        }

One last note, you can't save 2D Int arrays, so I flatten out the arrays and save them as "MapCoordinates"

    [DataContract]
    public class MapCoordinates
    {
        [DataMember]
        public int Row;
        [DataMember]
        public int Column;
        [DataMember]
        public Vector2 Location;

        public MapCoordinates(int x, int y) : this(x, y, Vector2.Zero) { }
        public MapCoordinates(int x, int y, Vector2 location)
        {
            Row = x;
            Column = y;
            Location = location;
        }

        public override string ToString()
        {
            return string.Format("Map Coordinates: {0}|{1} | Position: {2}", Row, Column, Location.ToString());
        }
    }

**Comment by craftworkgames** [comment]
@cupstudio Sorry I didn't comment on this sooner. I've been on holidays. 

The code looks pretty reasonable. I'm sure it'll be useful when we can around to sorting this out in MonoGame.Extended. Thanks for posting it.

We should definitely do some profiling. I'm really not sure at this stage how many tiles our tile renderer can currently handle.

**Comment by cupstudio** [comment]
Hi Dylan @craftworkgames ,

I have created a new Draw method for TiledTileLayer which only draws viewable tiles.  It uses the camera in order to do this, but could also be changed to use a more generic Rectangle.  

I used the existing Draw(Camera2D camera) method marked as Obsolete since it seemed to make sense for this example.

Under normal circumstances, the current Draw method will begin to drop the FPS after a few thousand tiles, depending on hardware.  I tested this version with 3 million tiles (3 tile layers, each at 1000x1000 - every tile filled in each layer) and the frame rate remains at 60 FPS solid, even while moving the camera around at a good clip.

You can see the changes in my fork, but the only change is this method, along with taking the Obsolete tag off of the base TileLayer class.

        public override void Draw(Camera2D camera)
        {
            var area = camera.GetBoundingRectangle();
            int firstCol = (int)Math.Floor(camera.Position.X / _map.TileWidth);
            int firstRow = (int)Math.Floor(camera.Position.Y / _map.TileHeight);

            // +3 to cover any gaps
            int numberOfColumns = (int)area.Width / _map.TileWidth + 3;
            int numberOfRows = (int)area.Height / _map.TileHeight + 3;

            _spriteBatch.Begin(transformMatrix: camera.GetViewMatrix(), blendState: BlendState.AlphaBlend, samplerState: SamplerState.PointClamp);
            int index;
            for (int row = firstRow; row <= firstRow + numberOfRows; row++)
            {
                index = row * _map.Width + firstCol;
                for (int x = 0; x < numberOfColumns; x++)
                {
                    // index must be greater than 0 and less than _tiles.Length
                    if (index >= 0 && index < _tiles.Length)
                    {
                        var region = _map.GetTileRegion(_tiles[index].Id);
                        if (region != null)
                        {
                            RenderLayer(_map, _tiles[index], region);
                        }
                    }
                    index++;
                }
            }
            _spriteBatch.End();
        }

**Comment by craftworkgames** [comment]
Fantasic! :smile: I'll take a closer look tomorrow and incorporate the changes into the library.

Do you want to do a pull request or shall I just copy and paste the code above?

**Comment by cupstudio** [comment]
I will do a pull request as I haven't done one before and would like to see the process.

Just a couple of quick thoughts around incorporating into the framework.

* Testing Isometric maps:  I haven't really done much with isometric maps or the isometric functionality in MGE (yet).  Need to test `case TiledMapOrientation.Isometric:` to ensure no compatibility issues with my `Draw` when it hits `RenderIsometric(TiledTile tile, TextureRegion2D region)`.  I don't expect an issue but I haven't closely examined how that works yet.
* Render Order:  I'll admit I couldn't think of the use case for this because I don't see what difference it would make as long as all tiles are rendered before the user has to see them.  My `Draw` doesn't use the `GetRenderOrderFunction` as a result.  Originally I was just going to make new or modified versions of the `IEnumerable<TiledTile> GetTilesXXXX` methods but I couldn't fathom how it would make a difference so I just do it Left-Down since that is the origin in the MonoGame framework.  Regardless it may be messy to have one `Draw` using them and another not?  Would like your thoughts on that.
* Lastly `TiledImageLayer` - Not sure if we would need a similar function here.

**Comment by craftworkgames** [comment]
> I will do a pull request as I haven't done one before and would like to see the process.

That's why I asked :wink: 

> Render Order: I'll admit I couldn't think of the use case for this because I don't see what difference it would make as long as all tiles are rendered before the user has to see them.

Render order doesn't make a lot of difference if your tile images are exactly the same size as the tile grid but I believe the Tiled map editor supports having tiles larger than the grid. This is when render order becomes important.

Here's the [Tiled github issue](https://github.com/bjorn/tiled/issues/455) with some screenshots explaining some use cases.

What I really need to do is setup some tests with various difference map types. I used to have some but for some silly reason I deleted them. I'll set this up after I merge your changes and fix it up.

> Lastly TiledImageLayer - Not sure if we would need a similar function here.

Maybe for very large images, but at this stage I wouldn't worry about it. Tiled supports image layers but tbh it's a pretty limited feature and I doubt anyone is actually using it.
